Demo/Program.cs

```diff
@@ -8,19 +9,35 @@ class Program
     {
         static void Main(string[] args)
         {
-            var videoEncoder = new VideoEncoder();
-            var audioEncoder = new AudioEncoder();
-
             var emailService = new EmailService();
             var smsService = new SmsService();
 
-            videoEncoder.Encoded += emailService.Send;
-            videoEncoder.Encoded += smsService.Send;
-            audioEncoder.Encoded += emailService.Send;
-            audioEncoder.Encoded += smsService.Send;
+            var notificationProvider = new NotificationProvider();
+            notificationProvider.Add(emailService);
+            notificationProvider.Add(smsService);
+
+            Console.WriteLine("Choose encoder type:\n a) Audio Encoder\n b) Video Encoder\n c) Integer Encoder\n Default encoder is AudioEncoder");
+
+            EncoderFactory encoderFactory = null;
+            var encoderType = Console.ReadLine();
+
+            encoderFactory = encoderType.ToLower() switch
+            {
+                "a" => new AudioFactory(),
+                "b" => new VideoFactory(),
+                "c" => new IntegerFactory(),
+                _ => new AudioFactory(),
```

Comment:
I think you could move these line into seperate extension method to keep Program.cs class cleaner